core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -972,6 +974,68 @@ protected Scope visitSetTableAuthorization(SetTableAuthorization node, Optional<
             return createAndAssignScope(node, scope);
         }
 
+        @Override
+        protected Scope visitTableExecute(TableExecute node, Optional<Scope> scope)
+        {
+            Table table = node.getTable();
+            QualifiedObjectName originalName = createQualifiedObjectName(session, table, table.getName());
+            if (metadata.getView(session, originalName).isPresent()) {
+                throw semanticException(NOT_SUPPORTED, node, "ALTER TABLE EXECUTE is not supported for views");
+            }
+
+            RedirectionAwareTableHandle redirection = metadata.getRedirectionAwareTableHandle(session, originalName);
+            QualifiedObjectName tableName = redirection.getRedirectedTableName().orElse(originalName);
+            TableHandle tableHandle = redirection.getTableHandle()
+                    .orElseThrow(() -> semanticException(TABLE_NOT_FOUND, table, "Table '%s' does not exist", tableName));
+
+            // TODO add access check
+            // accessControl.check...(session.toSecurityContext(), tableName);
+
+            // TODO Maybe make check optional; does it sometimes makes sense to execute table procedure on table with row filter or column masks?
+            if (!accessControl.getRowFilters(session.toSecurityContext(), tableName).isEmpty()) {
+                throw semanticException(NOT_SUPPORTED, node, "ALTER TABLE EXECUTE is not supported for table with ROW filter");
+            }
+
+            TableMetadata tableMetadata = metadata.getTableMetadata(session, tableHandle);
+            for (ColumnMetadata tableColumn : tableMetadata.getColumns()) {
+                if (!accessControl.getColumnMasks(session.toSecurityContext(), tableName, tableColumn.getName(), tableColumn.getType()).isEmpty()) {
+                    throw semanticException(NOT_SUPPORTED, node, "ALTER TABLE EXECUTE is not supported for table with column masks");
+                }
+            }
+
+            Scope tableScope = analyze(table, scope);
+
+            CatalogName catalogName = getRequiredCatalogHandle(metadata, session, node, tableName.getCatalogName());
+            String procedureName = node.getProcedureName().getValue();
+            TableProcedureMetadata procedureMetadata = metadata.getTableProcedureRegistry().resolve(catalogName, procedureName);
+
+            // analyze WHERE
+            if (!procedureMetadata.getExecutionMode().isFilter() && node.getWhere().isPresent()) {
+                throw semanticException(NOT_SUPPORTED, node, "WHERE not supported for procedure " + procedureName);
+            }
+            node.getWhere().ifPresent(where -> analyzeWhere(node, tableScope, where));
+
+            // analyze ORDER BY
```

Comment:
ORDER BY is nice because it allows us to tap into distributed sort when rewriting data
however, plan order by may be too limiting

- total ordering may be not required, sorting subsets of data can be sufficient (per file, per certain amount of data, grouped execution)
- different ordering schemes (eg z-order) may come useful. we should think how we will model them (an expression?)

Reply:
It is not used so far. And for local ordering (e.g for Z-ordering) we can express intention via `WITH` parameters.
e.g
```
WITH (z_order_columns = ARRAY['a', 'b']
```
Given that maybe we should drop support for `ORDER BY` for now?